kafka-admin/src/main/java/org/bf2/admin/security/TrustedSSOCallerPrincipalFactoryProducer.java

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+package org.bf2.admin.security;
+
+import io.smallrye.jwt.auth.principal.DefaultJWTCallerPrincipal;
+import io.smallrye.jwt.auth.principal.DefaultJWTCallerPrincipalFactory;
+import io.smallrye.jwt.auth.principal.DefaultJWTTokenParser;
+import io.smallrye.jwt.auth.principal.JWTAuthContextInfo;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipal;
+import io.smallrye.jwt.auth.principal.JWTCallerPrincipalFactory;
+import io.smallrye.jwt.auth.principal.KeyLocationResolver;
+import io.smallrye.jwt.auth.principal.ParseException;
+import org.bf2.admin.kafka.admin.KafkaAdminConfigRetriever;
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.jose4j.http.Get;
+import org.jose4j.jwk.HttpsJwks;
+import org.jose4j.jwt.consumer.JwtContext;
+import org.jose4j.keys.resolvers.VerificationKeyResolver;
+import org.jose4j.lang.UnresolvableKeyException;
+
+import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Alternative;
+import javax.enterprise.inject.Produces;
+import javax.interceptor.Interceptor;
+
+import java.util.logging.Logger;
+
+/**
+ * CDI producer to provide instance of {@code JWTCallerPrincipalFactory} capable
+ * of using a trusted certificate for JWKS retrieval.
+ *
+ * @deprecated needed until https://github.com/smallrye/smallrye-jwt/issues/571
+ *             is fixed and released
+ */
+@Deprecated(forRemoval = true)
+@ApplicationScoped
+@Alternative
+@Priority(Interceptor.Priority.APPLICATION + 10)
+public class TrustedSSOCallerPrincipalFactoryProducer {
+
+    private static final Logger LOG = Logger.getLogger(TrustedSSOCallerPrincipalFactoryProducer.class.getName());
+
+    private JWTCallerPrincipalFactory factory;
+
+    public TrustedSSOCallerPrincipalFactoryProducer() {
+        initCallerPrincipalFactory();
+    }
+
+    @Produces
+    public JWTCallerPrincipalFactory getFactory() {
+        return factory;
+    }
+
+    private void initCallerPrincipalFactory() {
+        LOG.info("Loading JWTCallerPrincipalFactory from " + getClass());
+        factory = new TrustedSSOCallerPrincipalFactory();
+        JWTCallerPrincipalFactory.setInstance(factory);
+    }
+
+    private static class TrustedSSOCallerPrincipalFactory extends DefaultJWTCallerPrincipalFactory {
+        private final DefaultJWTTokenParser parser = new TrustedSSOTokenParser();
+
+        @Override
+        public JWTCallerPrincipal parse(final String token, final JWTAuthContextInfo authContextInfo) throws ParseException {
+            JwtContext jwtContext = parser.parse(token, authContextInfo);
+            String type = jwtContext.getJoseObjects().get(0).getHeader("typ");
```

Comment:
I do wonder if this should be more defensive what happens if there are no `JoseObject` entries etc, but I see the upstream version does the same.

So its either an upstream issue or defended against else where...

Reply:
The call to `parser.parse` would result in an exception being thrown without a `JoseObject` present.